### ISSUE TITLE
Refactor beacon attestation validation

### ISSUE THREAD

**Issue by wemeetagain** [ChainSafe/lodestar#6446]
Should we move this into a function somewhere and be re-used in gossip + here + state transition? In a way to signal that this logic must be consistent between those three places

_Originally posted by @dapplion in https://github.com/ChainSafe/lodestar/pull/6198#discussion_r1432922813_

We should.

**Comment by HiroyukiNaito** [comment]
I pick the issue as a good first issue.

**Comment by HiroyukiNaito** [comment]
I'm locating where  should I change.

>Should we move this into a function somewhere and be re-used in gossip + here + state transition? In a way to signal that this logic must be consistent between those three places

Is this right place?
https://github.com/ChainSafe/lodestar/blob/unstable/packages/beacon-node/src/chain/validation/attestation.ts#L478

**Comment by HiroyukiNaito** [comment]
Shamely, I could not find where should I change the state-transition code.
Do you mind if you give me some hints?

**Comment by ensi321** [comment]
@HiroyukiNaito Hey! Thanks for picking this up.

Just want to share some background about this issue. Feel free to skip to tldr. 

The related change is for implementing Deneb EIP [EIP-7045](https://eips.ethereum.org/EIPS/eip-7045), which is to extend the admissible range of attestation from [n - 32, n] slots to [(Math.floor(n / 32) - 1) * 32, n] slots. You can check out our implementation #5731 and also the consensus spec [here](https://github.com/ethereum/consensus-specs/pull/3360).

The attestation validation logic in p2p we want to refactoring in Lodestar is to limit to lower bound of the slot in attestation by calculating `earliestPermissableSlot`. However, during state-transition, specifically during `process_attestation`, there is already an assertion in the [spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/beacon-chain.md?plain=1#L319) and [lodestar](https://github.com/ChainSafe/lodestar/blob/unstable/packages/state-transition/src/block/processAttestationPhase0.ts#L75) to limit the attestation to be within current or previous epoch, which has the same effect as setting lower (and upper) bound of the attestation slot. Therefore, the attestation validation logic in p2p is not needed in state-transition. 

> Shamely, I could not find where should I change the state-transition code.
Do you mind if you give me some hints?

tldr; The original comment is not quite accurate about state-transition. No refactoring needs to be made for state-transition. Only `beacon_aggregate_and_proof` and `beacon_attestation_{subnet_id}`


> Is this right place?
https://github.com/ChainSafe/lodestar/blob/unstable/packages/beacon-node/src/chain/validation/attestation.ts#L478

Yes. Should be a straightforward change. Let me know if you have further questions

**Comment by HiroyukiNaito** [comment]
@ensi321 Thank you for describing the technical context of this.
So, I try to crop the codes and functionalize them, and apply the 2 points you suggested.

**Comment by ail3ngrimaldi** [comment]
Hey! Was this finally solved? I would like to take a look. Any more insights? Since I have just seen that the discussion is a bit old.

**Comment by philknows** [comment]
Assigned to you @ail3ngrimaldi . I don't believe anything has been done to advance this since the conversation above... cc: @ensi321 ?

**Comment by ail3ngrimaldi** [comment]
Hey @philknows, Im trying to get where the code must be refactored, as I understood so far, the code has already been updated on state-transition, and we are trying to summit that logic in _someplace_ , so that gossip and attestation validation can also implement the [Deneb EIP-7045](https://eips.ethereum.org/EIPS/eip-7045). My questions so far:

1. Gossip and attestation validation still do not have implemented the extension of the admissible range of attestation?
2. Which part of this project structure would be suitable to upload this reusable function?

**Comment by ensi321** [comment]
> Hey [@philknows](https://github.com/philknows), Im trying to get where the code must be refactored, as I understood so far, the code has already been updated on state-transition, and we are trying to summit that logic in _someplace_ , so that gossip and attestation validation can also implement the [Deneb EIP-7045](https://eips.ethereum.org/EIPS/eip-7045). My questions so far:
> 
> 1. Gossip and attestation validation still do not have implemented the extension of the admissible range of attestation?
> 2. Which part of this project structure would be suitable to upload this reusable function?

Hi @ail3ngrimaldi , thanks for taking a look! The current state of our code is working correctly and I believe #7711 has already addressed the refactoring issue.

Closing this issue as it is been solved. Feel free to look at other issues
